client/txn.go

```diff
@@ -71,5 +174,28 @@ func (t *Txn) Prepare(calls ...Call) {
 func (t *Txn) Flush() error {
 	calls := t.prepared
 	t.prepared = nil
+	if len(calls) == 0 {
+		return nil
+	}
 	return t.kv.Run(calls...)
 }
+
+func (t *Txn) updateReadOnly(calls []Call) {
+	for _, c := range calls {
+		if b, ok := c.Args.(*proto.BatchRequest); ok {
+			for _, br := range b.Requests {
+				t.updateReadOnlyRequest(br.GetValue().(proto.Request))
+			}
+			continue
+		}
+		t.updateReadOnlyRequest(c.Args)
+	}
+}
+
+func (t *Txn) updateReadOnlyRequest(r proto.Request) {
+	if t.readOnly {
+		t.readOnly = !proto.IsTransactional(r)
```

Comment:
The naming of `IsTransactional` predates me. Agreed that the current name is confusing. I'll figure out something clearer for a subsequent change.